module/VuFind/src/VuFind/Search/SearchRunner.php

```diff
@@ -140,6 +140,10 @@ public function run(
             compact('params', 'request', 'runningSearchId')
         );
 
+        if (isset($rawRequest['enabledFacets'])) {
+            $params->retainFacets($rawRequest['enabledFacets']);
```

Comment:
While I don't see any major harm in globally enabling an enabledFacets query parameter, perhaps it would a more targeted fix to put this logic in the setup callback created by GetSideFacets::getFacetResults() instead of the SearchRunner itself.

Reply:
I tried to put this logic to setup callback at first, but it seems to be overridden later in the code, so I moved it to SearchRunner.

Reply:
Or maybe I am doing something wrong.

Reply:
I think what's happening is that the setup function is setting up the RecommendListener, which is then setting up the Facets recommendation module when the "setup complete" event triggers. So you're clearing out the facets first, then the recommendation module is putting them back in. I wonder if a better thing to do would be to manipulate the settings here:

https://github.com/vufind-org/vufind/blob/dev/module/VuFind/src/VuFind/AjaxHandler/GetSideFacets.php#L199-L203

Right now, there's already code that converts SideFacetsDeferred into SideFacets. Maybe we also need to adjust other configuration settings.

I admit that I can't remember all the details of this, so maybe that's not practical -- I realize that the recommendation module points to a configuration file section, so maybe that's too indirect to usefully manipulate. But I'm not sure I understand why SideFacetsDeferred would get set up in the configuration file with one set of facets, and then we would need to load a different set when the AJAX call occurs.

I wonder if @EreMaijala can shed any light on this upon his return.... I do think the existing solution here will probably work, but I wonder if we're just masking a problem that could be solved in a better way.

Reply:
One problem with the current patch, as far as I can see, is that it will prevent displaying counts for checkbox facets (we have a local implementation for that). So at least those should be taken into account.

In the bigger picture I think the problem is that side facets are initialized by RecommendListener and it all happens quite far away from the context of the search request, so passing information from the context is not so straight-forward. One option would be to bypass RecommendListener in GetSideFacets and just activate the required facets by providing a filter to the init method. I don't think it's feasible to modify the configuration beforehand since it's quite complex with some logic built into the SideFacets recommendation module. This of course breaks the isolation, but would also avoid enabling any extra recommenders when we just need results for side facets.

Reply:
I didn't review the code extremely closely, so maybe I'm missing something, but if the only reason we're using RecommendListener and the SideFacets module is to activate facets in the Params object, it does seem a lot more straightforward and efficient to just activate the facets we want directly in the setup callback instead of manipulating configurations and wiring up listeners. Then we can just turn on what we need, instead of turning everything on and then turning some things back off again.

It looks to me like there may be some cases where the AJAX routine is passed a list of facet fields (which is easy to deal with) and others where it's not. I didn't trace things back to see what all the circumstances might be. But one option would be to use the listener in cases where we don't know what we want, and to bypass it when we have a known list of facet fields.

Reply:
We still need to know something about the facet configuration, such as the number of facets to be displayed, and handling all that separately from SideFacets seems error-prone.

Reply:
Perhaps, yes. I'm a bit hesitant on making that modifiable by an AJAX request, though, since it would make it possible manually craft a request to e.g. ask for a million facets, and that might have serious performance implications.

Reply:
Unless I'm missing something, facet limits are initialized by the Params object, not by the recommendation module (https://github.com/vufind-org/vufind/blob/dev/module/VuFind/src/VuFind/Search/Params/FacetLimitTrait.php). I don't think SideFacets has anything to do with that.

Reply:
I think the piece that _is_ handled by SideFacets is figuring out which facets are OR facets and which are not... but I think there's less danger of abuse if those kinds of settings are passed in from the outside.

Reply:
Oh, sorry about the confusion. That wouldn't be scary at all then.

Reply:
Just reviving this thread to try to figure out where we left this -- I'm not sure who is waiting for whom at this point. I see that #2261 (discussed above) has now been merged, so maybe one possible next step is to add some tests to try to cover the changed functionality. @xrosecky, do you have any sense of what next steps should be? Do you need any specific guidance from @EreMaijala or me? (I think I'll have to take a bit of time to refresh my memory of this whole conversation before I can make a useful contribution... but I can at least tell at a glance that we never fully resolved it).